src/main/scala/nl/knaw/dans/easy/multideposit/Action.scala

```diff
@@ -15,32 +15,155 @@
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
 
 /**
-  * An action to be performed by Process SIP. It provides three methods that can be invoked to verify
-  * the feasibility of the action, to perform the action and - if necessary - to roll back the action.
-  */
-trait Action {
+ * An action to be performed by Process SIP. It provides three methods that can be invoked to verify
+ * the feasibility of the action, to perform the action and - if necessary - to roll back the action.
+ */
+/*
+  To future developers: this class is a semigroup. It satisfies the associativity law as defined
+  in https://wiki.haskell.org/Typeclassopedia#Laws_4.
+ */
+trait Action extends DebugEnhancedLogging {
+
+  private def logPreconditions(): Unit = {
+    logger.info(s"Checking preconditions of ${getClass.getSimpleName} ...")
+  }
+  private def logExecute(): Unit = {
+    logger.info(s"Executing action of ${getClass.getSimpleName} ...")
+  }
+  private def logRollback(): Unit = {
+    logger.info(s"An error occurred. Rolling back action ${getClass.getSimpleName} ...")
+  }
+  private[Action] def innerCheckPreconditions: Try[Unit] = Try(logPreconditions()).flatMap(_ => checkPreconditions)
+  private[Action] def innerExecute(): Try[Unit] = Try(logExecute()).flatMap(_ => execute())
+  private[Action] def innerRollback(): Try[Unit] = Try(logRollback()).flatMap(_ => rollback())
+
+  /**
+   * Verifies whether all preconditions are met for this specific action.
+   *
+   * @return `Success` when all preconditions are met, `Failure` otherwise
+   */
+  protected def checkPreconditions: Try[Unit] = Success(())
+
+  /**
+   * Exectue the action.
+   *
+   * @return `Success` if the execution was successful, `Failure` otherwise
+   */
+  protected def execute(): Try[Unit]
 
   /**
-    * Verifies whether all preconditions are met for this specific action.
-    *
-    * @return `Success` when all preconditions are met, `Failure` otherwise
-    */
-  def checkPreconditions: Try[Unit] = Success(())
+   * Cleans up results of a previous call to run so that a new call to run will not fail because of those results.
+   *
+   * @return `Success` if the rollback was successful, `Failure` otherwise
+   */
+  protected def rollback(): Try[Unit] = Success(())
 
   /**
-    * Exectue the action.
-    *
-    * @return `Success` if the execution was successfull, `Failure` otherwise
-    */
-  def run(): Try[Unit]
+   * Run an action. First the precondition is checked. If it fails a `PreconditionsFailedException`
+   * with a report is returned. Else, if the precondition succeed, the action is executed.
+   * If this fails, the action is rolled back and a `ActionRunFailedException` with a report is returned.
+   * If the execution was successful, `Success` is returned
+   *
+   * @return `Success` if the full execution was successful, `Failure` otherwise
+   */
+  final def run: Try[Unit] = {
```

Comment:
looks better indeed. I also changed `t` to `throwable`; seemed a better description...